qendpoint-store/src/main/java/com/the_qa_company/qendpoint/store/EndpointStoreQueryPreparer.java

```diff
@@ -128,15 +128,17 @@ protected CloseableIteration<? extends BindingSet, QueryEvaluationException> eva
 			new DisjunctiveConstraintOptimizer().optimize(tupleExpr, dataset, bindings);
 			new SameTermFilterOptimizer().optimize(tupleExpr, dataset, bindings);
 			new QueryModelNormalizerOptimizer().optimize(tupleExpr, dataset, bindings);
-			new QueryJoinOptimizer(evaluationStatistics).optimize(tupleExpr, dataset, bindings);
+			new QueryJoinOptimizer(evaluationStatistics, tripleSource).optimize(tupleExpr, dataset, bindings);
 			new IterativeEvaluationOptimizer().optimize(tupleExpr, dataset, bindings);
 			new FilterOptimizer().optimize(tupleExpr, dataset, bindings);
 			new OrderLimitOptimizer().optimize(tupleExpr, dataset, bindings);
 		}
 
 		new ServiceClauseOptimizer().optimize(tupleExpr, dataset, bindings);
 
-		return strategy.evaluate(tupleExpr, bindings);
+		QueryEvaluationStep precompile = strategy.precompile(tupleExpr);
+
+		return precompile.evaluate(bindings);
```

Comment:
Here is also a small performance fix that uses the precompile stage in RDF4J. I think this was introduced in RDF4J 4.0.0.